src/python/pants/source/wrapped_globs.py

```diff
@@ -4,37 +4,45 @@
 import os
 from abc import ABC, abstractmethod
 from hashlib import sha1
+from typing import Callable, Iterable, Iterator, List, Sequence, Tuple, Union, cast
 
 from twitter.common.dirutil.fileset import Fileset
+from typing_extensions import TypedDict
 
 from pants.base.build_environment import get_buildroot
-from pants.engine.fs import EMPTY_SNAPSHOT
+from pants.base.parse_context import ParseContext
+from pants.engine.fs import EMPTY_SNAPSHOT, Snapshot
 from pants.util.dirutil import fast_relpath, fast_relpath_optional
 from pants.util.memo import memoized_property
 
 
+class _GlobsDict(TypedDict, total=True):
+  globs: List[str]
+
+
+class Filespec(_GlobsDict, total=False):
```

Comment:
This is a little weird. The inheritance means that the dict must have all the fields from `_GlobsDict`, and `total=False` means that it may also optionally have the field `exclude`. The end effect is that that example in the docstring is properly typed.